Assets/Scripts/Scheduler/ScheduledEvent.cs

```diff
@@ -0,0 +1,175 @@
+        /// <param name="repeatsForever">Whether the event repeats forever (defaults false). If true repeats is ignored.</param>
+        /// <param name="repeats">Number of repeats (default 1). Ignored if repeatsForever=true.</param>
+        public ScheduledEvent(string name, Action<ScheduledEvent> onFire, float cooldown, bool repeatsForever = false, int repeats = 1)
+        {
+            this.Name = name;
+            this.OnFire = onFire;
+            this.Cooldown = cooldown;
+            this.TimeToWait = cooldown;
+            this.RepeatsForever = repeatsForever;
+            this.RepeatsLeft = repeats;
+            this.EventType = EventType.CSharp;
+        }
+
+        public ScheduledEvent(ScheduledEvent other)
+        {
+            this.Name = other.Name;
+            this.OnFire = other.OnFire;
+            this.Cooldown = other.Cooldown;
+            this.TimeToWait = other.Cooldown;
+            this.RepeatsForever = other.RepeatsForever;
+            this.RepeatsLeft = other.RepeatsLeft;
+            this.EventType = other.EventType;
+        }
+
+        public ScheduledEvent(EventPrototype eventPrototype, float cooldown, float timeToWait, bool repeatsForever = false, int repeats = 1)
+        {
+            this.Name = eventPrototype.Name;
+            this.OnFire = eventPrototype.OnFireCallback();
+            this.Cooldown = cooldown;
+            this.TimeToWait = timeToWait;
+            this.RepeatsForever = repeatsForever;
+            this.RepeatsLeft = repeats;
+            this.EventType = eventPrototype.EventType;
+        }
+
+        private event Action<ScheduledEvent> OnFire;
+
+        public string Name { get; protected set; }
+
+        public EventType EventType { get; protected set; }
+
+        public int RepeatsLeft { get; protected set; }
+
+        public float Cooldown { get; protected set; }
+
+        public float TimeToWait { get; protected set; }
+
+        public bool RepeatsForever { get; protected set; }
+
+        public bool LastShot
+        {
+            get
+            {
+                return RepeatsLeft == 1 && RepeatsForever == false;
+            }
+        }
+
+        public bool Finished
+        {
+            get
+            {
+                return RepeatsLeft < 1 && RepeatsForever == false;
+            }
+        }
+
+        /// <summary>
+        /// Advance the event clock by the specified deltaTime, and if it drops less that or equal to zero fire the event, resetting the clock to Cooldown.
+        /// Note: This fires the event multiple times if deltaTime is >= 2 * cooldown.
+        /// </summary>
+        /// <param name="deltaTime">Delta time in seconds (note: game time, not real time).</param>
+        public void Update(float deltaTime)
+        {
+            this.TimeToWait -= deltaTime;
+
+            while (this.TimeToWait <= 0)
+            {
+                Fire();
+                this.TimeToWait += this.Cooldown;
+            }
```

Comment:
The idea is to handle the case where deltaTime is large compared to the Cooldown of the event, so the event is scheduled to happen multiple times in the tick. The code above just makes sure the event fires when scheduled. This will probably be such a rare case as to not matter in practice unless there is a lag spike or an out of control mod (for example), in which case there are bigger problems going on than event scheduling.

So this is just handling an unlikely corner case. I could see an argument for ignoring the issue, and if people really want that logic gone I'll get rid of it.